### ISSUE TITLE
Dehumidifier (FGAC7044U1) No Longer Available after Update

### ISSUE THREAD

**Issue by sirjmann92** [bm1549/home-assistant-frigidaire#49]
I have a Frigidaire FGAC7044U1 dehumidifier that is no longer detected after the recent update to correct the login issues/API changes. The dehumidifier is displayed in the Frigidaire app, connected to the same WiFi, and verified working properly.

My Frigidaire Portable AC is detected and working, but the dehumidifier is no longer available in HA.

Thank you for the quick update for the auth issue!

**Comment by ha-elguano** [comment]
I have the Frigidaire FGAC5044W1 and am facing the same thing. 

I noticed in @rothn's revisions there is the following comment in humidifier.py:

>  _LOGGER.warning("Humidifier support temporarily disabled due to an API change and "
>                     "lack of test hardware")
>     """
>     Setting the humidity on the humidifier will raise a FrigidaireException explaining
>     the feature is not implemented because @rothn does not have one of their
>     humidifiers to test on. For the same reason, it's also possible that the humidifier
>     Destination is not correctly detected.
> 
>     These issues should be somewhat easily solvable by someone who has a compatible
>     humidifier manually calling `get_appliance_details()` on a humidifier, inspecting the
>     output, then making and testing changes based on climate.py. This can be easily
>     achieved with the frigidaire API in a Jupyter notebook.
> 
>     async_add_entities(
>         [
>             FrigidaireDehumidifier(client, appliance)
>             for appliance in appliances
>             if appliance.destination == frigidaire.Destination.DEHUMIDIFIER
>         ],
>         update_before_add=True,
>     )
>"""

My layman's read of this is that the current fix doesn't account for dehumidifiers or humidifiers. But it seems there is a somewhat straightforward solution given the work needed to get it working on the ACs!

**Comment by rothn** [comment]
Yes! If you're willing to do some Python work, I can send a Jupyter notebook for you to run that will capture the information I need to fix this for you. I'll then help you test the changes before we get them merged into main.

Some context: Last week, Frigidaire turned down the API that this plugin used (ECP) in favor of a new one (OCP) in the U.S. Previously, the maintainer temporarily worked around the issue in June 2023 by switching regions from LATAM to US: https://github.com/bm1549/frigidaire/commit/3b8e3541f69eb1ad878d18dc5a4b070702ad8534.

Because the changes were significant, I had to make minor changes to a dependency of this repo to account for the different level and format of information exposed by the new API. Information from people such as yourself can help get this working on more devices-- I only had a window AC unit available for testing.

If you are on Matrix, I can be reached for chat at @nicholasroth:nicholasroth.net so we can coordinate more easily. Otherwise, please let me know here if you would be willing to help.

**Comment by sarahemm** [comment]
I'm facing the same issue with a slightly different dehumidifier model, I'm not on Matrix but happy to help any way I can!

**Comment by rothn** [comment]
1. Open this Colab notebook: https://colab.research.google.com/drive/1T7C-u8sN5P_WW1dnS6rIv4mrP-Gp48xY#scrollTo=H8TLlFsI0gsx
2. Modify the "username" and "password" field to contain your Frigidaire username (email) and password.
![image](https://github.com/bm1549/home-assistant-frigidaire/assets/4665783/25e9d442-f37e-4334-9bc3-fea821d16c0f)
3. Run each Colab cell by clicking Runtime -> Run All.
![image](https://github.com/bm1549/home-assistant-frigidaire/assets/4665783/a35e57bc-5069-435a-8612-a1c80a8a5469)
You should see something like this:
![image](https://github.com/bm1549/home-assistant-frigidaire/assets/4665783/62cb5ed6-aa13-44cf-9d1b-75daaa526631)
4. Copy the output of the last cell into your clipboard and paste it into Github, removing any sensitive details
![image](https://github.com/bm1549/home-assistant-frigidaire/assets/4665783/322c2e94-f348-4d46-8c38-e84780c0e60a)

Here's what mine looks like. Please post yours in the comments. Notice how I've replaced my appliance ID with "<REDACTED>":
```
======= APPLIANCE 0  =======
==============================
###APPLIANCE CLASS STATE###
appliance_id: <REDACTED>
appliance_type: AC
destination: AC
nickname: Window AC
###APPLIANCE DETAILS###
{
    "cleanAirMode": "OFF",
    "sleepMode": "OFF",
    "applianceInfo": {
        "applianceType": "AC"
    },
    "temperatureRepresentation": "FAHRENHEIT",
    "applianceUiSwVersion": "v3",
    "uiLockMode": false,
    "targetTemperatureF": 72.0,
    "targetTemperatureC": 22.22222222222222,
    "applianceState": "OFF",
    "mode": "OFF",
    "alerts": [],
    "ambientTemperatureF": 50.0,
    "filterState": "CHANGE",
    "fanSpeedSetting": "AUTO",
    "networkInterface": {
        "swVersion": "v3.0.0_argo",
        "otaState": "IDLE",
        "linkQualityIndicator": "EXCELLENT",
        "niuSwUpdateCurrentDescription": "A21023902A-S00007029A",
        "swAncAndRevision": "S00007029A"
    },
    "verticalSwing": "OFF",
    "fanSpeedState": "LOW",
    "ambientTemperatureC": 10.0,
    "connectivityState": "connected"
}
###APPLIANCE CAPABILITIES###
{
    "alerts": {
        "access": "read",
        "type": "alert",
        "values": {
            "BUCKET_FULL": {},
            "BUS_HIGH_VOLTAGE": {},
            "COMMUNICATION_FAULT": {},
            "DC_MOTOR_FAULT": {},
            "DC_MOTOR_LOST_SPEED": {},
            "DRAIN_PAN_FULL": {},
            "INDOOR_DEFROST_THERMISTOR_FAULT": {},
            "PM25_SENSOR_FAULT": {},
            "TUBE_HIGH_TEMPERATURE": {},
            "UNKNOWN_STATE_ERROR": {}
        }
    },
    "ambientTemperatureC": {
        "access": "read",
        "type": "temperature"
    },
    "ambientTemperatureF": {
        "access": "read",
        "type": "temperature"
    },
    "applianceState": {
        "access": "read",
        "type": "string",
        "values": {
            "OFF": {},
            "RUNNING": {}
        }
    },
    "applianceUiSwVersion": {
        "access": "read",
        "type": "string"
    },
    "executeCommand": {
        "access": "write",
        "schedulable": true,
        "type": "string",
        "values": {
            "OFF": {},
            "ON": {}
        }
    },
    "fanSpeedSetting": {
        "access": "readwrite",
        "schedulable": true,
        "type": "string",
        "values": {
            "AUTO": {},
            "HIGH": {},
            "LOW": {},
            "MIDDLE": {}
        }
    },
    "fanSpeedState": {
        "access": "read",
        "type": "string",
        "values": {
            "HIGH": {},
            "LOW": {},
            "MIDDLE": {}
        }
    },
    "filterState": {
        "access": "read",
        "type": "string",
        "values": {
            "BUY": {},
            "CHANGE": {},
            "CLEAN": {},
            "GOOD": {}
        }
    },
    "mode": {
        "access": "readwrite",
        "schedulable": true,
        "triggers": [
            {
                "action": {
                    "fanSpeedSetting": {
                        "access": "readwrite",
                        "values": {
                            "HIGH": {},
                            "LOW": {},
                            "MIDDLE": {}
                        }
                    }
                },
                "condition": {
                    "operand_1": "value",
                    "operand_2": "FANONLY",
                    "operator": "eq"
                }
            },
            {
                "action": {
                    "fanSpeedSetting": {
                        "access": "readwrite",
                        "values": {
                            "AUTO": {},
                            "HIGH": {},
                            "LOW": {},
                            "MIDDLE": {}
                        }
                    }
                },
                "condition": {
                    "operand_1": "value",
                    "operand_2": "FANONLY",
                    "operator": "ne"
                }
            },
            {
                "action": {
                    "targetTemperatureC": {
                        "disabled": true,
                        "type": "temperature"
                    },
                    "targetTemperatureF": {
                        "disabled": true,
                        "type": "temperature"
                    }
                },
                "condition": {
                    "operand_1": {
                        "operand_1": "value",
                        "operand_2": "FANONLY",
                        "operator": "eq"
                    },
                    "operand_2": {
                        "operand_1": "value",
                        "operand_2": "5",
                        "operator": "eq"
                    },
                    "operator": "or"
                }
            },
            {
                "action": {
                    "targetTemperatureC": {
                        "disabled": false,
                        "type": "temperature"
                    },
                    "targetTemperatureF": {
                        "disabled": false,
                        "type": "temperature"
                    }
                },
                "condition": {
                    "operand_1": {
                        "operand_1": "value",
                        "operand_2": "FANONLY",
                        "operator": "ne"
                    },
                    "operand_2": {
                        "operand_1": "value",
                        "operand_2": "5",
                        "operator": "ne"
                    },
                    "operator": "and"
                }
            },
            {
                "action": {
                    "fanSpeedSetting": {
                        "disabled": true
                    }
                },
                "condition": {
                    "operand_1": "value",
                    "operand_2": "6",
                    "operator": "eq"
                }
            },
            {
                "action": {
                    "fanSpeedSetting": {
                        "disabled": false
                    }
                },
                "condition": {
                    "operand_1": "value",
                    "operand_2": "6",
                    "operator": "ne"
                }
            },
            {
                "action": {
                    "sleepMode": {
                        "access": "read"
                    }
                },
                "condition": {
                    "operand_1": {
                        "operand_1": "value",
                        "operand_2": "FANONLY",
                        "operator": "eq"
                    },
                    "operand_2": {
                        "operand_1": "value",
                        "operand_2": "5",
                        "operator": "eq"
                    },
                    "operator": "or"
                }
            },
            {
                "action": {
                    "sleepMode": {
                        "access": "readwrite"
                    }
                },
                "condition": {
                    "operand_1": {
                        "operand_1": "value",
                        "operand_2": "FANONLY",
                        "operator": "ne"
                    },
                    "operand_2": {
                        "operand_1": "value",
                        "operand_2": "5",
                        "operator": "ne"
                    },
                    "operator": "and"
                }
            }
        ],
        "type": "string",
        "values": {
            "COOL": {},
            "ECO": {},
            "FANONLY": {},
            "OFF": {
                "disabled": true
            }
        }
    },
    "networkInterface": {
        "command": {
            "access": "write",
            "type": "string",
            "values": {
                "APPLIANCE_AUTHORIZE": {},
                "START": {},
                "USER_AUTHORIZE": {},
                "USER_NOT_AUTHORIZE": {}
            }
        },
        "linkQualityIndicator": {
            "access": "read",
            "type": "string",
            "values": {
                "EXCELLENT": {},
                "GOOD": {},
                "POOR": {},
                "UNDEFINED": {},
                "VERY_GOOD": {},
                "VERY_POOR": {}
            }
        },
        "niuSwUpdateCurrentDescription": {
            "access": "read",
            "type": "string"
        },
        "otaState": {
            "access": "read",
            "type": "string",
            "values": {
                "DESCRIPTION_AVAILABLE": {},
                "DESCRIPTION_DOWNLOADING": {},
                "DESCRIPTION_READY": {},
                "FW_DOWNLOADING": {},
                "FW_DOWNLOAD_START": {},
                "FW_SIGNATURE_CHECK": {},
                "FW_UPDATE_IN_PROGRESS": {},
                "IDLE": {},
                "READY_TO_UPDATE": {},
                "UPDATE_ABORT": {},
                "UPDATE_CHECK": {},
                "UPDATE_ERROR": {},
                "UPDATE_OK": {},
                "WAITINGFORAUTHORIZATION": {}
            }
        },
        "startUpCommand": {
            "access": "write",
            "type": "string",
            "values": {
                "UNINSTALL": {}
            }
        },
        "swAncAndRevision": {
            "access": "read",
            "type": "string"
        },
        "swVersion": {
            "access": "read",
            "type": "string"
        }
    },
    "sleepMode": {
        "access": "readwrite",
        "type": "string",
        "values": {
            "OFF": {},
            "ON": {}
        }
    },
    "targetTemperatureC": {
        "access": "readwrite",
        "default": 15.56,
        "max": 32.22,
        "min": 15.56,
        "schedulable": true,
        "step": 1.0,
        "type": "temperature"
    },
    "targetTemperatureF": {
        "access": "readwrite",
        "default": 60.0,
        "max": 90.0,
        "min": 60.0,
        "schedulable": true,
        "step": 1.0,
        "type": "temperature"
    },
    "temperatureRepresentation": {
        "access": "readwrite",
        "type": "string",
        "values": {
            "CELSIUS": {},
            "FAHRENHEIT": {}
        }
    },
    "uiLockMode": {
        "access": "readwrite",
        "type": "boolean",
        "values": {
            "OFF": {},
            "ON": {}
        }
    },
    "verticalSwing": {
        "access": "readwrite",
        "type": "string",
        "values": {
            "OFF": {},
            "ON": {}
        }
    }
}
```

**Comment by rothn** [comment]
@sarahemm @ha-elguano @boctok @sirjmann92

Since I don't have a dehumidifier, I need your help completing the above steps to make dehumidifiers work with the new API. Thanks! The data from just one person should probably be enough, but more is better. Just be mindful of your privacy, and don't post anything you don't feel comfortable sharing.

**Comment by ha-elguano** [comment]
@rothn , thank you for the detailed instructions here, as well as your willingness to help! I'm happy to jump in to help capture data as well.

QQ - how does running the collab cells gather the data? Is it because we have dehumidifiers registered with Frigidaire Cloud, and this will pull the necessary data from the Cloud responses to the Collab calls?

In other words, there's no need for an actual networked connection to the dehumidifier hardware, correct?

**Comment by rothn** [comment]
Changes are ready to test. The testing repo has a self-contained frigidaire module, since adding back humidifier support requires changes both to this plugin and the API implementation (separate repo).

I've tested these commands, and everything seems to work for my window AC. To test, run the following:
```
git clone -b fix-dehumidifiers-testing https://github.com/rothn/home-assistant-frigidaire
cd home-assistant-frigidaire
git submodule init
git submodule update
```

Next, copy the contents of the `custom_components` directory to your HomeAssistant `custom_components`, as described in the repository README.

If it works, great-- please say so! If not, please post logs here.

**Comment by rothn** [comment]
Tagging @sarahemm @ha-elguano @boctok @sirjmann92 for testing :-)

**Comment by ha-elguano** [comment]
@rothn Where would I run the git command? It isn't recognized in HA console  or proxmox shell.

**Comment by ha-elguano** [comment]
@rothn I figured it out the cloning. Followed the instructions, re-added the integration in HA, and so far, it works!

- My dehumidifier shows up as available again. 
- It correctly displays status as on and off in my HA Dash (with a short delay between changes, just as before). 
- I can adjust the humidity in the HA entity and it mostly works (there is a "snap-to" in 5% increments, so you have to change it fast enough to snap to the next value before the command is sent)
- I can change fan modes from HA entity

So far, everything seems to work great!

My hat is off to you, @rothn!!!

**Comment by sirjmann92** [comment]
@rothn - thanks for the quick update, once again!

With your updates, my dehumidifier (FGAC7044U1) is showing up again. I clicked around and changed states and it seems to function without error or issue. All states are available, reported status is correct, and I'm able to make changes successfully. Obviously I don't have any real test-cases to follow and more thorough usage is needed to really say that it's 100% working but it seems to be working fine from the brief testing I did.

If anything pops up, I'll open a separate issue.

I really appreciate your effort!

**Comment by sarahemm** [comment]
Can confirm this seems to fix mine (model FGAC5044W1) as well! Let me know if providing the Colab notebook output is still useful and I can take care of it later today, or if you have enough info already. Thanks so much for your work on this, I really appreciate it. I've had ours switching modes from Dry to Quiet overnight automatically via HA since we got it and it'll be nice to have that functionality back.

**Comment by rothn** [comment]
It's good to hear that this appears to work. I've created https://github.com/bm1549/frigidaire/pull/24 to add dehumidifier support back to the API package. Once that is merged and releases, I'll remove the API subpackage from the testing branch and submit a PR to this repository that will hopefully be suitable to merge.

**Comment by justinmyersm** [comment]
> It's good to hear that this appears to work. I've created [bm1549/frigidaire#24](https://github.com/bm1549/frigidaire/pull/24) to add dehumidifier support back to the API package. Once that is merged and releases, I'll remove the API subpackage from the testing branch and submit a PR to this repository that will hopefully be suitable to merge.

Will this also fix the issue with Window AC units?

**Comment by rothn** [comment]
@justinmyersm Window ACs should already work after updating to the latest version of the plugin.

**Comment by rothn** [comment]
I've cherry-picked the fix into a new branch/MR: https://github.com/bm1549/home-assistant-frigidaire/pull/50/files

It's ready to merge as soon as @bm1549 cuts a new release for PYPI and I test everything again to make sure there are no packaging surprises. In the meantime, folks should be able to directly use the testing branch.

**Comment by rothn** [comment]
@justinmyersm Would you mind following the instructions above with the Colab and filing a new bug against https://github.com/bm1549/frigidaire with the output? This looks suspiciously like an API-related issue.

**Comment by rothn** [comment]
For anyone still using the testing branch while we wait for changes to go upstream, the new testing branch is called "in-tree-frigidaire-for-testing" and can be pulled like:
```
git clone -b in-tree-frigidaire-for-testing https://github.com/rothn/home-assistant-frigidaire
cd home-assistant-frigidaire
git submodule init
git submodule update
```

Remember to copy the contents of the custom_components directory to your HomeAssistant custom_components, as described in the repository README. Sorry for the delay getting this to work upstream! We will have to wait for the "frigidaire" module to release before this will be available through HACS and the main branch.

**Comment by rothn** [comment]
Yet another update: A release has been cut and https://github.com/bm1549/home-assistant-frigidaire/pull/50 was merged. You all should be able to run the `main` branch now or use HACS-- whatever the original installation method was.

**Comment by rothn** [comment]
@sirjmann92 Ready to close the issue?

**Comment by jhemak** [comment]
Fixed it for me - thank you!!  But, did we lose the compressor_running and fan_running attributes?

**Comment by rothn** [comment]
The new API does not expose compressor information based on the data in my appliance or the others I have seen here. Fan information is still exposed through fan modes, though the OFF state now comes from the appliance mode rather than the fan itself.

**Comment by jhemak** [comment]
Ah, that's too bad!  I don't see anything now to show if the unit is actually running.  The state of the entity shows if it's powered on or off, and the attributes seem to reflect the specified settings rather than current operation.

**Comment by rothn** [comment]
Yes, I don't remember anything that would indicate this in the output. If you would like to try, consider checking out applianceState in more details. That may have it. You could also cycle through different states and see what changes in the output to see what's available. But kind of OT for this issue.

**Comment by sirjmann92** [comment]
@rothn yep! All good! Thanks again for the help and support!

**Comment by marty22877** [comment]
> > @justinmyersm Window ACs should already work after updating to the latest version of the plugin.
> 
> I am just now getting time to update, and I am not able to get it to work. Initially I just copied the contents of the updated files and rebooted HA. That didn't work, so I deleted the folder, rebooted, redownloaded, rebooted, and I am getting an error when trying to log in. I only have 1 window AC unit and I logged out and back into my Frigidaire app just to make sure I was using the right login. What am I doing wrong?
> 
> ![image](https://private-user-images.githubusercontent.com/12924428/302400023-0f7f0c0c-3894-4f60-9512-eb103f97d55f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ0MDYzNTQsIm5iZiI6MTcxNDQwNjA1NCwicGF0aCI6Ii8xMjkyNDQyOC8zMDI0MDAwMjMtMGY3ZjBjMGMtMzg5NC00ZjYwLTk1MTItZWIxMDNmOTdkNTVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI5VDE1NTQxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUzMjVjZjY3MWI4MWY0MWJiZmViZTBmNThkYjIxN2RmY2Q1MWIyMmQyYjU4YWI0ODVjNzJlNTEwNTM4OWE1MGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.75lH4dDDOWzxdqmx0QHl7yDAWpM4QfV6hEj5-gI9eqo)
> 
> ```
> This error originated from a custom integration.
> 
> Logger: custom_components.frigidaire.config_flow
> Source: custom_components/frigidaire/config_flow.py:66
> Integration: frigidaire
> First occurred: 1:42:13 PM (3 occurrences)
> Last logged: 1:44:57 PM
> 
> Unexpected exception
> Traceback (most recent call last):
>   File "/config/custom_components/frigidaire/config_flow.py", line 66, in async_step_user
>     await validate_input(self.hass, user_input)
>   File "/config/custom_components/frigidaire/config_flow.py", line 38, in validate_input
>     appliances = await hass.async_add_executor_job(
>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
>     result = self.fn(*self.args, **self.kwargs)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/config/custom_components/frigidaire/config_flow.py", line 31, in setup
>     return client.get_appliances()
>            ^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/frigidaire/__init__.py", line 367, in get_appliances
>     return get_appliances_inner()
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/frigidaire/__init__.py", line 364, in get_appliances_inner
>     return list(map(generate_appliance, appliances))
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/frigidaire/__init__.py", line 353, in generate_appliance
>     return Appliance(raw_appliance)
>            ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/frigidaire/__init__.py", line 82, in __init__
>     self.appliance_type: str = args['properties']['reported']['applianceInfo']['applianceType']
>                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
> KeyError: 'applianceInfo'
> ```

@justinmyersm did you ever get a fix to this?
